### PR TITLE
PR for code review

### PR DESCRIPTION
**Pull request by Samuelimza**
Initial PR.

### CODE REVIEW

File: GA2/Car.pde [hunk 1]
```diff
@@ -0,0 +1,230 @@
+class Car {
```
Comment (AarushJ) [inline GA2/Car.pde]: add access specifier

File: GA2/Car.pde [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+  //physics attributes
+  boolean dead = false;
+  boolean acc = false, left = false, right = false, decc = false;
+  PVector pos;
+  PVector vel;
+  float drag = 0.96;
```
Comment (AarushJ) [inline GA2/Car.pde]: use final if these are not to be used later

File: GA2/Car.pde [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    } else if (species == 4) {
+      // 20%
+      mutationRate = 0.20;
+    } else if (species == 5) {
+      // 100%
+      mutationRate = 1.0;
+    }
+  }
+
+  void update() {
```
Comment (AarushJ) [inline GA2/Car.pde]: change name to update_car_state()

File: GA2/Car.pde [hunk 4]
```diff
@@ -0,0 +1,230 @@
+class Car {
+  //physics attributes
+  boolean dead = false;
+  boolean acc = false, left = false, right = false, decc = false;
```
Comment (AarushJ) [inline GA2/Car.pde]: acc -> acceleration

Comment (AarushJ) [reply GA2/Car.pde]: decc -> deacceleration

File: GA2/Car.pde [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+  //Genetic attributes
+  int proximitySensorLength = 200;
+  int species;
+  boolean isBest = false;
+  float fitness = 0;
+  float mutationRate = 0.01;
+  int previousMarkerIndex = -1;
+
+  Car(int x, int y, int species) {
```
Comment (AarushJ) [inline GA2/Car.pde]: use final in parameters also if their value is not changing later.

File: GA2/Car.pde [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+    pos = new PVector(x, y);
+    vel = new PVector(0, 0);
+    this.species = species;
+    proximity = new float[3][1];
+    neuralNetwork = new NNetwork();
+    if (species == 1) {
```
Comment (AarushJ) [inline GA2/Car.pde]: use switch case here

File: GA2/Car.pde [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+      //if controlled by neural network and colliding with walls then die
+      if (notOnTrack()) {
+        dead = true;
+      }
+    }
+  }
+
+  void updateMarkerStatus() {
+    for (int i = 0; i < markers.size(); i++) {
+      Marker current = markers.get(i);
```
Comment (AarushJ) [inline GA2/Car.pde]: current -> currentMarkerIndex

File: GA2/Car.pde [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+  //check for points near the periphery of the car that lie in the wall for collision detection
+  boolean notOnTrack() {
+    int index0 = (int)pos.x + 7 + (int)pos.y * mapWidth;
+    int index1 = (int)pos.x - 7 + (int)pos.y * mapWidth;
+    int index2 = (int)pos.x + (int)(pos.y + 7) * mapWidth;
+    int index3 = (int)pos.x + (int)(pos.y - 7) * mapWidth;
+    boolean condition0 = index0 < 0 || index1 < 0 || index2 < 0 || index3 < 0;
+    boolean condition1 = index0 >= myMap.pixels.length || index1 >= myMap.pixels.length || index2 >= myMap.pixels.length || index3 >= myMap.pixels.length;
+    if (!condition0 && !condition1) {
+      if (myMap.pixels[index0] == -16777216 || myMap.pixels[index1] == -16777216 || myMap.pixels[index2] == -16777216 || myMap.pixels[index3] == -16777216) {
```
Comment (AarushJ) [inline GA2/Car.pde]: avoid magic numbers. Store such values in variables and then use those variables everywhere needed .

File: GA2/Car.pde [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+  }
+
+  //drawSensors uses cars angle heading to draw proximity sensors and pass arguments to the proximity findDistance function.
+  void drawSensors() {
+    PVector heading = PVector.fromAngle(angle - PI / 6);
+    heading.mult(proximitySensorLength);
+    stroke(0, 0, 0, 100);
+    //drawing for the -30 degree sensor
+    if (debugMode) {
+
```
Comment (AarushJ) [inline GA2/Car.pde]: remove this line

File: GA2/Car.pde [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+    //drawing for the 0 degree sensor
+    heading.rotate(-PI / 6);
+    if (debugMode) {
+      line(pos.x, pos.y, pos.x + heading.x, pos.y + heading.y);
+    }
+    findDistance(heading.copy(), 1);
+    //println(proximity[0][0] + ", " + proximity[1][0] + ", " + proximity[2][0]);
+  }
+
+  void show() {
```
Comment (AarushJ) [inline GA2/Car.pde]: change function name to something clear

File: GA2/Car.pde [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+    if (!(!debugMode && dead)) {
+      stroke(0);
+      pushMatrix();
+      if (species == 1) {
```
Comment (AarushJ) [inline GA2/Car.pde]: switch case

File: GA2/GA2.pde [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+void setup() {
+  size(1056, 600);
+  textAlign(LEFT, TOP);
+  //create font object of arial with 25 size
+  f = createFont("Arial", 25, true);
+  textFont(f);
+  rectMode(CENTER);
+  noStroke();
+  myMap = loadImage("data/tracks/track" + track + ".png");
```
Comment (AarushJ) [inline GA2/GA2.pde]: Read about MessageFormat class and use message formatting here.

File: GA2/GA2.pde [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+  mapWidth = myMap.width;
+  mapHeight = myMap.height;
+  println(mapWidth + ", " + mapHeight);
+  myMap.loadPixels();
+  LeftView = loadImage("data/LeftView.png");
```
Comment (AarushJ) [inline GA2/GA2.pde]: constant here for string

File: GA2/GA2.pde [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+  for (int i = 0; i < cars.length; i++) {
+    int species = (i / (noOfCars / 5)) + 1;
+    cars[i] = new Car(370, 30, species);
+    if(track == 3){
+      cars[i].pos.y = startY[1];
+    }
+  }
+  markers = new ArrayList<Marker>();
+  loadMarkers("D:/NewFolder/Osama/Programming/Java/Processing/GA2/data");
```
Comment (AarushJ) [inline GA2/GA2.pde]: constant here for string

File: GA2/GA2.pde [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+  //background(myMap);
+  //translate(228, 0);
+  image(LeftView, 600, 0);//-228, 0);
+  image(myMap, 0, 0);
+  //cp5.draw();
+  stroke(4);
+  line(828, 0, 828, 600);
+  fill(187, 252, 184, 150);
+  if (!testing) {
+    text("Gen: " + ga.generation, 0, 0);
```
Comment (AarushJ) [inline GA2/GA2.pde]: use message formattter here also

File: GA2/GA2.pde [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+
+public void LoadBest() {
+  String name = cp5.get(Textfield.class, "loadName").getText();
+  ga.loadGeneration(name);
+  testing = true;
+  ga.setTestingCar();
+}
+
+void loadMarkers(String path) {
+  JSONArray markersArray = loadJSONArray(path + "/markers1.json");
```
Comment (AarushJ) [inline GA2/GA2.pde]: use Message formatter here also

File: GA2/GeneticAlgorithm.pde [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+      println("Reproduced due to No Activity");
+    }
+  }
+
+  void reproduce() {
+    maxFitness = 0;
+    for (int i = 0; i < cars.length; i++) {
+      if (cars[i].fitness > maxFitness) {
+        maxFitness = cars[i].fitness;
+        maxFitI = i;
```
Comment (AarushJ) [inline GA2/GeneticAlgorithm.pde]: change to maxFitnessIndex

File: GA2/GeneticAlgorithm.pde [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+    } else if (lastActivitySecond < second - 3) {
+      reproduce();
+      counter = 0;
+      println("Reproduced due to No Activity");
+    }
+  }
+
+  void reproduce() {
+    maxFitness = 0;
+    for (int i = 0; i < cars.length; i++) {
```
Comment (AarushJ) [inline GA2/GeneticAlgorithm.pde:48]: make separate method for calculating max and max index (follow Single Responsibility Principle)

File: GA2/GeneticAlgorithm.pde [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+        testingCar.neuralNetwork = new NNetwork(cars[bestCarFindingIndex].neuralNetwork);
+        testingCar.isBest = true;
+        break;
+      }
+    }
+  }
+
+  void saveGeneration(Car[] cars, String saveName) {
+    JSONArray generationArray = new JSONArray();
+    JSONObject metaData = new JSONObject();
```
Comment (AarushJ) [inline GA2/GeneticAlgorithm.pde:159]: create separate method for metadata generation.

File: GA2/GeneticAlgorithm.pde [hunk 20]
```diff
@@ -0,0 +1,225 @@
...
+          for (int k = 0; k < cars[carCounter].neuralNetwork.weights[i][j].length; k++) {
+            carJSON.setFloat(str(weightCounter), cars[carCounter].neuralNetwork.weights[i][j][k]);
+            weightCounter++;
+          }
+        }
+      }
+      carJSON.setBoolean("isBest", cars[carCounter].isBest);
+      generationArray.setJSONObject(carCounter + 1, carJSON);
+    }
+    saveJSONArray(generationArray, "generations/metaDataGens/" + saveName + ".json");
```
Comment (AarushJ) [inline GA2/GeneticAlgorithm.pde]: Use message formatter here

File: GA2/GeneticAlgorithm.pde [hunk 21]
```diff
@@ -0,0 +1,225 @@
...
+  }
+
+  void loadGeneration(String generationToLoad) {
+    JSONArray generationArray = loadJSONArray("generations/metaDataGens/" + generationToLoad + ".json");
```
Comment (AarushJ) [inline GA2/GeneticAlgorithm.pde]: Use message formatter here

File: GA2/GeneticAlgorithm.pde [hunk 22]
```diff
@@ -0,0 +1,225 @@
...
+    try {
+      generation = metaData.getInt("Generation");
+      tempDrag = metaData.getFloat("drag");
+      tempAngularDrag = metaData.getFloat("angularDrag");
+      tempPower = metaData.getFloat("power");
+      tempTurnSpeed = metaData.getFloat("turnSpeed");
+      tempBraking = metaData.getFloat("braking");
+      tempProximitySensorLength = metaData.getInt("proximitySensorLength");
+    }
+    catch(Exception exception) {
```
Comment (AarushJ) [inline GA2/GeneticAlgorithm.pde]: 1. Avoid generic exception. Write exact exception. 
2. If there are multiple exceptions which u think my occur, use multiple catch statements with the single try block above. 
3. Learn to use logs . See @Slf4j annotation of lombok.

File: GA2/Markers.pde [hunk 23]
```diff
@@ -0,0 +1,27 @@
+class Marker{
```
Comment (AarushJ) [inline GA2/Markers.pde]: @AllArgsConstructor or something similar in lombok. use it here.

File: GA2/NNetwork.pde [hunk 24]
```diff
@@ -0,0 +1,51 @@
+class NNetwork{
+  int[] architecture = {3, 5, 4, 4};
```
Comment (AarushJ) [inline GA2/NNetwork.pde]: change architecture to a more obvious name

**Review by AarushJ — CHANGES_REQUESTED** [review]

File: GA2/Car.pde (hunk 2 above)
Comment (Samuelimza) [reply GA2/Car.pde]: while loading a save, they need to be changed, hence cannot be set final.

File: GA2/Car.pde (hunk 3 above)
Comment (Samuelimza) [reply GA2/Car.pde]: Done.

File: GA2/Car.pde (hunk 4 above)
Comment (Samuelimza) [reply GA2/Car.pde]: Done.

File: GA2/Car.pde (hunk 5 above)
Comment (Samuelimza) [reply GA2/Car.pde]: Done.

File: GA2/Car.pde (hunk 6 above)
Comment (Samuelimza) [reply GA2/Car.pde]: Done.

File: GA2/Car.pde (hunk 7 above)
Comment (Samuelimza) [reply GA2/Car.pde]: Done.

File: GA2/Car.pde (hunk 9 above)
Comment (Samuelimza) [reply GA2/Car.pde]: Done.

File: GA2/Car.pde (hunk 8 above)
Comment (Samuelimza) [reply GA2/Car.pde]: Done.

File: GA2/Car.pde (hunk 10 above)
Comment (Samuelimza) [reply GA2/Car.pde]: Done.

File: GA2/Car.pde (hunk 11 above)
Comment (Samuelimza) [reply GA2/Car.pde]: Done.

File: GA2/GA2.pde (hunk 13 above)
Comment (Samuelimza) [reply GA2/GA2.pde]: Done.

File: GA2/GA2.pde (hunk 14 above)
Comment (Samuelimza) [reply GA2/GA2.pde]: Done.

File: GA2/GeneticAlgorithm.pde (hunk 17 above)
Comment (Samuelimza) [reply GA2/GeneticAlgorithm.pde]: Done.

File: GA2/GeneticAlgorithm.pde (hunk 18 above)
Comment (Samuelimza) [reply GA2/GeneticAlgorithm.pde:48]: Done.

File: GA2/GeneticAlgorithm.pde (hunk 19 above)
Comment (Samuelimza) [reply GA2/GeneticAlgorithm.pde:159]: Done.

File: GA2/NNetwork.pde (hunk 24 above)
Comment (Samuelimza) [reply GA2/NNetwork.pde]: Done.

File: GA2/Car.pde (hunk 1 above)
Comment (Samuelimza) [reply GA2/Car.pde]: Done.

File: GA2/GA2.pde (hunk 12 above)
Comment (Samuelimza) [reply GA2/GA2.pde]: Done.

File: GA2/GA2.pde (hunk 15 above)
Comment (Samuelimza) [reply GA2/GA2.pde]: Done.

File: GA2/GA2.pde (hunk 16 above)
Comment (Samuelimza) [reply GA2/GA2.pde]: Path was constant hence set as final instead.

File: GA2/GeneticAlgorithm.pde (hunk 21 above)
Comment (Samuelimza) [reply GA2/GeneticAlgorithm.pde]: Done.

File: GA2/GeneticAlgorithm.pde (hunk 20 above)
Comment (Samuelimza) [reply GA2/GeneticAlgorithm.pde]: Done.

File: GA2/GeneticAlgorithm.pde (hunk 22 above)
Comment (Samuelimza) [reply GA2/GeneticAlgorithm.pde]: Done 1, 2.
Looking into 3.

**Pull request merged**